cirq/contrib/qasm_import/_parser.py

```diff
@@ -35,7 +36,56 @@ def __init__(self, supported_format: bool, qelib1_include: bool,
         self.circuit = c
 
 
-class QasmParser(object):
+class QasmGateStatement:
+
+    def __init__(self, qasm_gate: str, cirq_gate: cirq.Gate, num_args: int):
+        self.qasm_gate = qasm_gate
+        self.cirq_gate = cirq_gate
+        self.num_args = num_args
+
+    def validate_args(self, args: List[List[cirq.Qid]], lineno: int):
+        if len(args) != self.num_args:
+            raise QasmException(
+                "{} only takes {} arg(s) (qubits and/or registers), "
+                "got: {}, at line {}".format(self.qasm_gate, self.num_args,
+                                             len(args), lineno))
+
+    def on(self, args: List[List[cirq.Qid]],
+           lineno: int) -> Iterable[cirq.Operation]:
+        self.validate_args(args, lineno)
+        reg_size = 1
+        for reg in args:
+            if len(reg) > 1 and len(reg) != reg_size:
+                if reg_size == 1:
+                    reg_size = len(reg)
+                else:
+                    raise QasmException(
+                        "Non matching quantum registers of length {} "
+                        "at line {}".format([len(reg) for reg in args], lineno))
+
+        # OpenQASM gates can be applied on single qubits and qubit registers.
+        # We represent single qubits as registers of size 1.
+        # Based on the OpenQASM spec (https://arxiv.org/abs/1707.03429),
+        # single qubit arguments can be mixed with qubit registers.
+        # Given quantum registers of length reg_size and single qubits are both
+        # used as arguments, we generate reg_size GateOperations via iterating
+        # through each qubit of the registers 0 to n-1 and use the same one
+        # qubit from the "single-qubit registers" for each operation.
+        for i in range(reg_size):
```

Comment:
I believe this whole loop is:

```
for qubits in functools.reduce(np.broadcast, reg):
    yield self.cirq_gate.on(*qubits)
```

But it may do the wrong thing for singleton lists.

Reply:
thank you for the suggestion - it totally works with an extra check for validation of uniqueness! Coming soon...